### PR TITLE
internal/dag: report certificate delegation failure correctly

### LINKED ISSUE #1347: TcpProxy IngressRoute status is not set when it references an un-delegated TLS cert
**What steps did you take and what happened:**
1. Create a secret in the `heptio-contour` namespace called `default-ssl-cert`.

2. Create an IngressRoute that references the secret:
```yaml
apiVersion: contour.heptio.com/v1beta1
kind: IngressRoute
metadata:
  name: sample-app
  namespace: tls-cert-delegation
spec:
  virtualhost:
    fqdn: 127.0.0.1.nip.io
    tls:
      secretName: heptio-contour/default-ssl-cert
  tcpproxy:
    services:
    - name: sample-app
      port: 80
```

3. Get the IngressRoute and see status:
```shell
$ kubectl get ingressroute
NAME         FQDN               TLS SECRET                        FIRST ROUTE   STATUS   STATUS DESCRIPTION
sample-app   127.0.0.1.nip.io   heptio-contour/default-ssl-cert
```

**What did you expect to happen:**
The status should reflect that the user does not have access to the referenced certificate. To avoid information disclosure, perhaps the message should be something like: "The referenced certificate does not exist, or you do not have access to it"

**Environment:**

- Contour version: v0.14.1
- Kubernetes version: (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.0", GitCommit:"641856db18352033a0d96dbc99153fa3b27298e5", GitTreeState:"clean", BuildDate:"2019-03-26T00:04:52Z", GoVersion:"go1.12.1", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.0", GitCommit:"e8462b5b5dc2584fdcd18e6bcfe9f1e4d970a529", GitTreeState:"clean", BuildDate:"2019-06-25T23:41:27Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}
```

### PR DESCRIPTION
**Pull request by davecheney**
Fixes #1347

Signed-off-by: Dave Cheney <dave@cheney.net>

### CODE REVIEW

**Comment by davecheney** [comment]
@alexbrand this is the current behaviour for #1347. Could you please review and let me know it differs from what you expect. Thank you.

**Comment by alexbrand** [comment]
My bad @davecheney. It looks like the issue only happens when the certificate exists in the other namespace. I missed that in the reproduction steps.

**Comment by davecheney** [comment]
@alexbrand i believe I have fixed the issue, PTAL.

File: internal/dag/builder.go [hunk 1]
```diff
@@ -354,11 +354,15 @@ func (b *Builder) computeIngressRoutes() {
...
-				svhost.MinProtoVersion = MinProtoVersion(ir.Spec.VirtualHost.TLS.MinimumProtocolVersion)
-				enforceTLS = true
+			if sec != nil {
+				if b.delegationPermitted(m, ir.Namespace) {
+					svhost := b.lookupSecureVirtualHost(host)
+					svhost.Secret = sec
+					svhost.MinProtoVersion = MinProtoVersion(ir.Spec.VirtualHost.TLS.MinimumProtocolVersion)
+					enforceTLS = true
+				} else {
+					b.setStatus(Status{Object: ir, Status: StatusInvalid, Description: fmt.Sprintf("%s: certificate delegation not permitted", tls.SecretName)})
```
Comment (alexbrand) [inline internal/dag/builder.go]: Do we need a `continue` after this, to stop processing the IngressRoute?

**Comment by alexbrand** [comment]
I believe this will also fix #1348

File: internal/dag/builder.go (hunk 1 above)
Comment (davecheney) [reply internal/dag/builder.go]: technically yes, but the next condition, `sec == nil` will prevent the other 1/2 of this function running. Having said that, this is a horrid piece of nested logic. Let me see if I can clean it up some so we can use `continue` or `return` after a status setting.

**Pull request merged**